pkg/pki/issue.go

```diff
@@ -0,0 +1,169 @@
+	// This func returns a cert, private key and a bool.  The bool value is whether the keypair is stored
+	// in a legacy format. This bool is used by a keypair
+	// task to convert a Legacy Keypair to the new Keypair API format.
+	FindKeypair(name string) (*Certificate, *PrivateKey, bool, error)
+}
+
+// IssueCert issues a certificate, either a self-signed CA or from a CA in a keystore.
+func IssueCert(request *IssueCertRequest, keystore Keystore) (issuedCertificate *Certificate, issuedKey *PrivateKey, caCertificate *Certificate, err error) {
+	certificateType := request.Type
+	if expanded, found := wellKnownCertificateTypes[certificateType]; found {
+		certificateType = expanded
+	}
+
+	template := &x509.Certificate{
+		BasicConstraintsValid: true,
+		IsCA:                  false,
+		SerialNumber:          request.Serial,
+	}
+
+	tokens := strings.Split(certificateType, ",")
+	for _, t := range tokens {
+		if strings.HasPrefix(t, "KeyUsage") {
+			ku, found := parseKeyUsage(t)
+			if !found {
+				return nil, nil, nil, fmt.Errorf("unrecognized certificate option: %v", t)
+			}
+			template.KeyUsage |= ku
+		} else if strings.HasPrefix(t, "ExtKeyUsage") {
+			ku, found := parseExtKeyUsage(t)
+			if !found {
+				return nil, nil, nil, fmt.Errorf("unrecognized certificate option: %v", t)
+			}
+			template.ExtKeyUsage = append(template.ExtKeyUsage, ku)
+		} else if t == "CA" {
+			template.IsCA = true
+		} else {
+			return nil, nil, nil, fmt.Errorf("unrecognized certificate option: %q", t)
+		}
+	}
+
+	template.Subject = request.Subject
+
+	var alternateNames []string
+	alternateNames = append(alternateNames, request.AlternateNames...)
+
+	for _, san := range alternateNames {
+		san = strings.TrimSpace(san)
+		if san == "" {
+			continue
+		}
+		if ip := net.ParseIP(san); ip != nil {
+			template.IPAddresses = append(template.IPAddresses, ip)
+		} else {
+			template.DNSNames = append(template.DNSNames, san)
+		}
+	}
+
+	var caPrivateKey *PrivateKey
+	var signer *x509.Certificate
+	if !template.IsCA {
+		var err error
+		caCertificate, caPrivateKey, _, err = keystore.FindKeypair(request.Signer)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if caPrivateKey == nil {
+			return nil, nil, nil, fmt.Errorf("ca key for %q was not found; cannot issue certificates", request.Signer)
+		}
+		if caCertificate == nil {
+			return nil, nil, nil, fmt.Errorf("ca certificate for %q was not found; cannot issue certificates", request.Signer)
+		}
+		signer = caCertificate.Certificate
+	}
+
+	privateKey := request.PrivateKey
+	if privateKey == nil {
+		var err error
+		privateKey, err = GeneratePrivateKey()
+		if err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
+	if request.MinValidDays != 0 {
+		hash := fnv.New32()
+		addrs, err := net.InterfaceAddrs()
+		sort.Slice(addrs, func(i, j int) bool {
+			return addrs[i].String() < addrs[j].String()
+		})
+		if err == nil {
```

Comment:
Nit: we should probably warn in this error case.  It might also be easier to understand if the random generation was split into its own function.  Finally, I do wonder if it's worth it, vs a simple random number?

Reply:
We want all certs generated on the same node to have the same validity period, so everything on the node breaks at the same time and there's more time until the next node breaks. A simple random number would have different things on the node breaking at different times and different nodes having components breaking at times closer to each other.

Reply:
Conflicted, so rebased and addressed those items.